### PR TITLE
Simple continous time system example

### PR DESCRIPTION
**Pull request by RussTedrake**
As documented in the code, this is meant to be a sort of "hello world" example for the drake::system classes.  It defines a very simple continous time system, simulates it from a given initial condition.  

I elected to make it <double> only and tried to make it minimal and readable.  Comments welcome.

TODO (in future PRs): 
 - would love to actually plot the result; needs #3734 and #3647 .
 - will update the textbook documentation to use the C++ examples!
 - delete the matlab version of this example.

---
This change is [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/3735)

### CODE REVIEW

File: drake/examples/simple_continuous_time_system.cc [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+                            1,  // dimension of output (y) = 1
+                            systems::kContinuousSampling);
+  };
+  ~SimpleContinuousTimeSystem() override{};
+
+  // xdot = -x + x^3
+  void EvalTimeDerivatives(
+      const systems::Context<double>& context,
+      systems::ContinuousState<double>* derivatives) const override {
+    double x = context.get_continuous_state()->get_state().GetAtIndex(0);
```
Comment (RussTedrake) [inline drake/examples/simple_continuous_time_system.cc]: would love to have one less level of redirection in here.  (e.g. context.get_continuous_state_vector().GetAtIndex(0)).  

also, perhaps rename ContinousState<T>.get_state to ContinuousState<T>.get_vector?

File: drake/examples/simple_continuous_time_system.cc [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+  void EvalOutput(const systems::Context<double>& context,
+                  systems::SystemOutput<double>* output) const override {
+    double x = context.get_continuous_state()->get_state().GetAtIndex(0);
+    output->GetMutableVectorData(0)->SetAtIndex(0, x);
+  }
+
+ protected:
+  // allocate a basic vector of dimension 1
+  virtual std::unique_ptr<systems::ContinuousState<double>>
+  AllocateContinuousState() const override {
```
Comment (RussTedrake) [inline drake/examples/simple_continuous_time_system.cc]: This method is probably the hardest for a newbie to appreciate.  And it's quite a different workflow than the way that the input/output ports are declared.  Perhaps we could have a class that provides a default implementation of this, and the number of states is declared (e.g. in the constructor)?

**Comment by david-german-tri** [comment]
:lgtm: Consider squashing the commits.

---

Reviewed 1 of 2 files at r1, 2 of 2 files at r2.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[drake/examples/simple_continuous_time_system.cc, line 32 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/3735#-KTewznPREY1yOE5HwNy-r1-32:-KTkOQeUA9IeQ9YhysdE:b449dy5) ([raw file](https://github.com/robotlocomotion/drake/blob/2c3c7d3c443a1808ceb670cd3a20e839dd535366/drake/examples/simple_continuous_time_system.cc#L32)):*
<details><summary><i>Previously, RussTedrake (Russ Tedrake) wrote…</i></summary>
> would love to have one less level of redirection in here.  (e.g. context.get_continuous_state_vector().GetAtIndex(0)).  
> 
> 
> 
> also, perhaps rename ContinousState<T>.get_state to ContinuousState<T>.get_vector?
> 
> 
</details>
I agree; this has been a recurring complaint.  #3751

---

*[drake/examples/simple_continuous_time_system.cc, line 47 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/3735#-KTewznPREY1yOE5HwNy-r1-47:-KTkOWpLw0KmwLggo8O1:bqsss2w) ([raw file](https://github.com/robotlocomotion/drake/blob/2c3c7d3c443a1808ceb670cd3a20e839dd535366/drake/examples/simple_continuous_time_system.cc#L47)):*
<details><summary><i>Previously, RussTedrake (Russ Tedrake) wrote…</i></summary>
> This method is probably the hardest for a newbie to appreciate.  And it's quite a different workflow than the way that the input/output ports are declared.  Perhaps we could have a class that provides a default implementation of this, and the number of states is declared (e.g. in the constructor)?
</details>
If "continuous non-second order state with no named accessors" is a common enough case to justify syntactic sugar, then yes, no reason why not.  #3751

---

*[drake/examples/simple_continuous_time_system.cc, line 48 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/3735#-KTkPfj4jI041Hxaaj54:-KTkPfj5UWuDldBlAfMF:b-7z6b09) ([raw file](https://github.com/robotlocomotion/drake/blob/9d51a094c8e82e93bf5a7a0c0a9135baeaa0071e/drake/examples/simple_continuous_time_system.cc#L48)):*
> ```cpp
>   std::unique_ptr<systems::ContinuousState<double>>
>   AllocateContinuousState() const override {
>     return std::make_unique<systems::ContinuousState<double>>(
> ```

BTW, we could also provide a `ContinuousState<T>(int size)` constructor that sugars this.

---

*[drake/examples/simple_continuous_time_system.cc, line 58 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/3735#-KTkPp1qCeekSrWVAVwY:-KTkPp1qCeekSrWVAVwZ:bb7o539) ([raw file](https://github.com/robotlocomotion/drake/blob/9d51a094c8e82e93bf5a7a0c0a9135baeaa0071e/drake/examples/simple_continuous_time_system.cc#L58)):*
> ```cpp
> int main(int argc, char* argv[]) {
>   // create the simple system
>   auto system = std::make_unique<drake::examples::SimpleContinuousTimeSystem>();
> ```

BTW, I think this could just be on the stack.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3735#-:-KTkQ3ZkMvNCZrD7aTMu:bjjhqly)*

**Comment by RussTedrake** [comment]
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/examples/simple_continuous_time_system.cc, line 48 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/3735#-KTkPfj4jI041Hxaaj54:-KTkT0CeS9TT93Bsuaok:brxr4nf) ([raw file](https://github.com/robotlocomotion/drake/blob/9d51a094c8e82e93bf5a7a0c0a9135baeaa0071e/drake/examples/simple_continuous_time_system.cc#L48)):*
<details><summary><i>Previously, david-german-tri (David German) wrote…</i></summary>
> BTW, we could also provide a `ContinuousState<T>(int size)` constructor that sugars this.
</details>
That could be useful, too.  But I'd love to not have to write this method for simple systems.

---

*[drake/examples/simple_continuous_time_system.cc, line 58 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/3735#-KTkPp1qCeekSrWVAVwY:-KTkT65ha8dfE0yAbtCZ:b-3o0hhz) ([raw file](https://github.com/robotlocomotion/drake/blob/9d51a094c8e82e93bf5a7a0c0a9135baeaa0071e/drake/examples/simple_continuous_time_system.cc#L58)):*
<details><summary><i>Previously, david-german-tri (David German) wrote…</i></summary>
> BTW, I think this could just be on the stack.
</details>
good call.  that will further simplify.  if you see anything else that could make the code better/simpler, please do let me know.  as I said, my hope is that it can be tutorial.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3735)*

**Comment by RussTedrake** [comment]
Review status: 2 of 3 files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/examples/simple_continuous_time_system.cc, line 58 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/3735#-KTkPp1qCeekSrWVAVwY:-KTkVUUpSqO6qXCv1grQ:b-katzo6) ([raw file](https://github.com/robotlocomotion/drake/blob/9d51a094c8e82e93bf5a7a0c0a9135baeaa0071e/drake/examples/simple_continuous_time_system.cc#L58)):*
<details><summary><i>Previously, RussTedrake (Russ Tedrake) wrote…</i></summary>
> good call.  that will further simplify.  if you see anything else that could make the code better/simpler, please do let me know.  as I said, my hope is that it can be tutorial.
</details>
it occurred to me that the tutorial value would also go up if I removed the namespaces (so every reference is explicit).  i don't think it's much less readable, since I was already using system:: a bunch, and just had to make it drake::system:: instead.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3735)*

**Comment by david-german-tri** [comment]
:lgtm:

---

Review status: 2 of 3 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3735#-:-KTkWhbEQEfT8ay2zv9j:bnfp4nl)*

**Comment by sherm1** [comment]
Reviewed 1 of 2 files at r1, 1 of 2 files at r2.
Review status: 2 of 3 files reviewed at latest revision, 5 unresolved discussions.

---

*[drake/examples/simple_continuous_time_system.cc, line 32 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/3735#-KTewznPREY1yOE5HwNy-r1-32:-KTkQ8K4QMzSAxhjLF-F:bw2ym4v) ([raw file](https://github.com/robotlocomotion/drake/blob/2c3c7d3c443a1808ceb670cd3a20e839dd535366/drake/examples/simple_continuous_time_system.cc#L32)):*
<details><summary><i>Previously, david-german-tri (David German) wrote…</i></summary>
> I agree; this has been a recurring complaint.  #3751
</details>
BTW, my personal preference would be `context.get_xc()` (returning an indexable vector) but apparently not everyone sees the beauty in that! I like it because it would be unambiguous and perfectly clear after looking at the documentation once. OTOH, `get_continuous_state()` is confusing no matter how many times I see it.

---

*[drake/examples/simple_continuous_time_system.cc, line 47 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/3735#-KTewznPREY1yOE5HwNy-r1-47:-KTkTspOR4SjTBwLMP-S:b65ilfm) ([raw file](https://github.com/robotlocomotion/drake/blob/2c3c7d3c443a1808ceb670cd3a20e839dd535366/drake/examples/simple_continuous_time_system.cc#L47)):*
<details><summary><i>Previously, david-german-tri (David German) wrote…</i></summary>
> If "continuous non-second order state with no named accessors" is a common enough case to justify syntactic sugar, then yes, no reason why not.  #3751
</details>
For newbies I agree it would be nice to hide this. But the real workflow is that the System class defines the topological pieces necessary to build a Diagram (basically input and output ports), then requests Context resources later, only after the complete Diagram is available. State variables (incl. continuous/difference/mode) are context resources and an advanced user can have a peek at inputs and parameter values before deciding how much, and what kinds, of State resources to request. (Cache resources are treated similarly.)

---

*[drake/examples/simple_continuous_time_system.cc, line 6 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/3735#-KTkPVSloRE3SQ1dwFhK:-KTkPVSloRE3SQ1dwFhL:bbfmf53) ([raw file](https://github.com/robotlocomotion/drake/blob/9d51a094c8e82e93bf5a7a0c0a9135baeaa0071e/drake/examples/simple_continuous_time_system.cc#L6)):*
> ```cpp
> /// This is meant to be a sort of "hello world" example for the
> /// drake::system classes.  It defines a very simple continous time system,
> /// simulates it from a given initial condition, and plots the result.
> ```

This comment is at file scope -- you could tell doxygen with `@file` but I think it would be more useful to move it down to the class.

---

*[drake/examples/simple_continuous_time_system.cc, line 67 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/3735#-KTkWupQZcPUFX1SCC8G:-KTkWupQZcPUFX1SCC8H:b-7219nb) ([raw file](https://github.com/robotlocomotion/drake/blob/9d51a094c8e82e93bf5a7a0c0a9135baeaa0071e/drake/examples/simple_continuous_time_system.cc#L67)):*
> ```cpp
>   initial_conditions->get_mutable_continuous_state()
>       ->get_mutable_state()
>       ->SetAtIndex(0, .9);
> ```

BTW, we should overload `operator[]` for indexing to make these nicer. How about:
```c++
initial_conditions->get_mutable_xc()[0] = 0.9;
```
or 
```c++
initial_conditions->get_mutable_xc(0) = 0.9;
```

---

*[drake/examples/simple_continuous_time_system.cc, line 76 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/3735#-KTkXXJPVwqnNd724JTn:-KTkXXJPVwqnNd724JTo:b-3izss4) ([raw file](https://github.com/robotlocomotion/drake/blob/9d51a094c8e82e93bf5a7a0c0a9135baeaa0071e/drake/examples/simple_continuous_time_system.cc#L76)):*
> ```cpp
>                             .get_continuous_state()
>                             ->CopyToVector()
>                             .value()) < 1.0e-4);
> ```

BTW, yikes!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3735)*

**Comment by sherm1** [comment]
Reviewed 1 of 2 files at r2.
Review status: 2 of 3 files reviewed at latest revision, 5 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3735)*

**Comment by sherm1** [comment]
:lgtm: except for a comment comment, PTAL.

---

Review status: 2 of 3 files reviewed at latest revision, 5 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3735#-:-KTkXqrS8X1h_iC4jD43:b-dy5zlx)*

**Comment by david-german-tri** [comment]
Review status: 2 of 3 files reviewed at latest revision, 5 unresolved discussions.

---

*[drake/examples/simple_continuous_time_system.cc, line 47 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/3735#-KTewznPREY1yOE5HwNy-r1-47:-KTkZVroiD4GYucbtxdA:bn1mcji) ([raw file](https://github.com/robotlocomotion/drake/blob/2c3c7d3c443a1808ceb670cd3a20e839dd535366/drake/examples/simple_continuous_time_system.cc#L47)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary>
> For newbies I agree it would be nice to hide this. But the real workflow is that the System class defines the topological pieces necessary to build a Diagram (basically input and output ports), then requests Context resources later, only after the complete Diagram is available. State variables (incl. continuous/difference/mode) are context resources and an advanced user can have a peek at inputs and parameter values before deciding how much, and what kinds, of State resources to request. (Cache resources are treated similarly.)
</details>
@sherm1 Yes, agreed, which is one of the reasons I describe this as sugar.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3735)*

**Comment by RussTedrake** [comment]
Review status: 2 of 3 files reviewed at latest revision, 5 unresolved discussions.

---

*[drake/examples/simple_continuous_time_system.cc, line 32 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/3735#-KTewznPREY1yOE5HwNy-r1-32:-KTkmSHDGEPxEEt_TnQA:b-jzd5n4) ([raw file](https://github.com/robotlocomotion/drake/blob/2c3c7d3c443a1808ceb670cd3a20e839dd535366/drake/examples/simple_continuous_time_system.cc#L32)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary>
> BTW, my personal preference would be `context.get_xc()` (returning an indexable vector) but apparently not everyone sees the beauty in that! I like it because it would be unambiguous and perfectly clear after looking at the documentation once. OTOH, `get_continuous_state()` is confusing no matter how many times I see it.
</details>
i used xc in my matlab, but the continuous_state is ok for me.  (the c in xc stands for continuous, so i don't see it being any better)

---

*[drake/examples/simple_continuous_time_system.cc, line 6 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/3735#-KTkPVSloRE3SQ1dwFhK:-KTkm4rvCSZuG0K5kiXH:b-3nresq) ([raw file](https://github.com/robotlocomotion/drake/blob/9d51a094c8e82e93bf5a7a0c0a9135baeaa0071e/drake/examples/simple_continuous_time_system.cc#L6)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary>
> This comment is at file scope -- you could tell doxygen with `@file` but I think it would be more useful to move it down to the class.
</details>
the class also has documentation, and this description refers to more than the class (it also mentions simulation of the class).  i've added the @file.

---

*[drake/examples/simple_continuous_time_system.cc, line 67 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/3735#-KTkWupQZcPUFX1SCC8G:-KTkmOHXGEgo2T8WVhvv:b-2fjtwv) ([raw file](https://github.com/robotlocomotion/drake/blob/9d51a094c8e82e93bf5a7a0c0a9135baeaa0071e/drake/examples/simple_continuous_time_system.cc#L67)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary>
> BTW, we should overload `operator[]` for indexing to make these nicer. How about:
> ```c++
> initial_conditions->get_mutable_xc()[0] = 0.9;
> ```
> or 
> ```c++
> initial_conditions->get_mutable_xc(0) = 0.9;
> ```
</details>
completely agree.  will add it to #3751.

---

*[drake/examples/simple_continuous_time_system.cc, line 76 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/3735#-KTkXXJPVwqnNd724JTn:-KTknT6CMBWczhDCM6bB:b6mi1sy) ([raw file](https://github.com/robotlocomotion/drake/blob/9d51a094c8e82e93bf5a7a0c0a9135baeaa0071e/drake/examples/simple_continuous_time_system.cc#L76)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary>
> BTW, yikes!
</details>
agreed.  yikes.  it's that or get_state.GetAtIndex(0).  i will switch to that to be more consistent with the usage above.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3735)*

**Comment by sherm1** [comment]
Reviewed 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[drake/examples/simple_continuous_time_system.cc, line 32 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/3735#-KTewznPREY1yOE5HwNy-r1-32:-KTl4lZ6j_hEoZaYB3tv:b-f85lpr) ([raw file](https://github.com/robotlocomotion/drake/blob/2c3c7d3c443a1808ceb670cd3a20e839dd535366/drake/examples/simple_continuous_time_system.cc#L32)):*
<details><summary><i>Previously, RussTedrake (Russ Tedrake) wrote…</i></summary>
> i used xc in my matlab, but the continuous_state is ok for me.  (the c in xc stands for continuous, so i don't see it being any better)
</details>
In my fantasy world, I picture this set of methods, each returning an indexable vector:
```c++
get_x()  // returns continuous and difference state variables together
get_xc() // where x={xc,xd}
get_xd()
get_q()  // where xc={q,v,z}
get_v()
get_z()
```
It seems awkward to invent long names for all of these when the math is so concise. But I'm OK with the long names :cry:.  

---

*[drake/examples/simple_continuous_time_system.cc, line 6 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/3735#-KTl4ZbHEu-MalFWStR_:-KTl4ZbHEu-MalFWStRa:bx8hr28) ([raw file](https://github.com/robotlocomotion/drake/blob/3d06ba8d816bc6add9904a78ad027e63c8af85a5/drake/examples/simple_continuous_time_system.cc#L6)):*
> ```cpp
> ///
> /// This is meant to be a sort of "hello world" example for the
> /// drake::system classes.  It defines a very simple continous time system,
> ```

BTW, continous -> continuous

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3735)*

**Comment by RussTedrake** [comment]
Review status: 2 of 3 files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/examples/simple_continuous_time_system.cc, line 32 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/3735#-KTewznPREY1yOE5HwNy-r1-32:-KTlKo2O2vrrqSYUWKIj:b91l4xf) ([raw file](https://github.com/robotlocomotion/drake/blob/2c3c7d3c443a1808ceb670cd3a20e839dd535366/drake/examples/simple_continuous_time_system.cc#L32)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary>
> In my fantasy world, I picture this set of methods, each returning an indexable vector:
> ```c++
> get_x()  // returns continuous and difference state variables together
> get_xc() // where x={xc,xd}
> get_xd()
> get_q()  // where xc={q,v,z}
> get_v()
> get_z()
> ```
> It seems awkward to invent long names for all of these when the math is so concise. But I'm OK with the long names :cry:.  
> 
</details>
that's almost exactly what we had in matlab.  i'm ok with either.

---

*[drake/examples/simple_continuous_time_system.cc, line 6 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/3735#-KTl4ZbHEu-MalFWStR_:-KTlKythIVk1CJ7uReoE:b-7rl0jd) ([raw file](https://github.com/robotlocomotion/drake/blob/3d06ba8d816bc6add9904a78ad027e63c8af85a5/drake/examples/simple_continuous_time_system.cc#L6)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary>
> BTW, continous -> continuous
</details>
done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3735)*

**Comment by jwnimmer-tri** [comment]
FTR:
- #3760 fixed a build error from this PR.
- The implementation comments in `main` have sentences that are not capitalized nor punctuated correctly.

**Comment by jwnimmer-tri** [comment]
Right.  The example's comments are discarded:
- http://www.stack.nl/~dimitri/doxygen/manual/config.html#cfg_extract_local_classes
- https://github.com/RobotLocomotion/drake/blob/06661beb39393934bbee7d0d787bea036ca5910d/drake/doc/Doxyfile_CXX.in#L68
It is defective to use doxygen-style comments when our doxygen configuration does not extract them.

**Comment by RussTedrake** [comment]
@jwnimmer-tri .  you're tough.  :)  
i'm actually of the opinion that we could set EXTRACT_LOCAL_CLASSES to YES.  I don't think there are many local classes around; the only reason that there is one here is for the tutorial value of having everything in one, simple to digest place.

**Comment by jwnimmer-tri** [comment]
I am good with turning on as much Doxygen as you like.  Personally, I think it should be on everywhere -- local classes, anonymous classes, private methods, test code, examples, everything!  I think "Uses Doxygen" and "Is a user-facing API" should not be assumed to be the same concept, but others may disagree.

**Comment by sherm1** [comment]
> but others may disagree

Me, for example!

**Comment by RussTedrake** [comment]
I will take a spin.  After hitting send it occurred to me that there might be a lot of small structs, etc, that would quickly pollute things.  I will likely convert this file's comments to non-doxygen instead.

**Pull request merged**